### PR TITLE
8319577: x86_64 AVX2 intrinsics for Arrays.sort methods (int, float arrays)

### PR DESCRIPTION
**Pull request by vamsi-parasa**
The goal is to develop faster sort routines for x86_64 CPUs by taking advantage of AVX2 instructions. This enhancement provides an order of magnitude speedup for Arrays.sort() using int, long, float and double arrays.

For serial sort on random data, this PR shows upto ~7.5x improvement for 32-bit datatypes (int, float) on Intel TigerLake machine as shown in the performance data below.

For parallel sort on random data, this PR shows upto ~3.4x for 32-bit datatypes (int, float) as shown below.

**Note:** This PR also improves the performance of AVX512 sort by upto 35%.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/sparasa/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/sparasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Benchmark (Serial Sort) | Size | Baseline      (us/op) | AVX2     (us/op) | Speedup
-- | -- | -- | -- | --
ArraysSort.intSort | 10 | 0.034 | 0.029 | 1.2
ArraysSort.intSort | 25 | 0.088 | 0.044 | 2.0
ArraysSort.intSort | 50 | 0.239 | 0.159 | 1.5
ArraysSort.intSort | 75 | 0.417 | 0.27 | 1.5
ArraysSort.intSort | 100 | 0.572 | 0.265 | 2.2
ArraysSort.intSort | 1000 | 10.098 | 4.282 | 2.4
ArraysSort.intSort | 10000 | 330.065 | 43.383 | 7.6
ArraysSort.intSort | 100000 | 4099.527 | 778.943 | 5.3
ArraysSort.intSort | 1000000 | 49150.16 | 9634.335 | 5.1
ArraysSort.floatSort | 10 | 0.045 | 0.043 | 1.0
ArraysSort.floatSort | 25 | 0.105 | 0.073 | 1.4
ArraysSort.floatSort | 50 | 0.278 | 0.216 | 1.3
ArraysSort.floatSort | 75 | 0.476 | 0.241 | 2.0
ArraysSort.floatSort | 100 | 0.583 | 0.313 | 1.9
ArraysSort.floatSort | 1000 | 10.182 | 4.329 | 2.4
ArraysSort.floatSort | 10000 | 323.136 | 57.175 | 5.7
ArraysSort.floatSort | 100000 | 4299.519 | 862.63 | 5.0
ArraysSort.floatSort | 1000000 | 50889.4 | 10972.19 | 4.6

</body>

</html>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/sparasa/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/sparasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Benchmark    (Parallel Sort) | Size | Baseline      (us/op) | AVX2     (us/op) | Speedup
-- | -- | -- | -- | --
ArraysSort.intParallelSort | 10 | 0.033 | 0.029 | 1.14
ArraysSort.intParallelSort | 25 | 0.087 | 0.043 | 2.02
ArraysSort.intParallelSort | 50 | 0.25 | 0.161 | 1.55
ArraysSort.intParallelSort | 75 | 0.419 | 0.267 | 1.57
ArraysSort.intParallelSort | 100 | 0.555 | 0.262 | 2.12
ArraysSort.intParallelSort | 1000 | 8.287 | 4.259 | 1.95
ArraysSort.intParallelSort | 10000 | 412.983 | 138.082 | 2.99
ArraysSort.intParallelSort | 100000 | 857.161 | 250.89 | 3.42
ArraysSort.intParallelSort | 1000000 | 7961.746 | 3090.108 | 2.58
ArraysSort.floatParallelSort | 10 | 0.047 | 0.043 | 1.09
ArraysSort.floatParallelSort | 25 | 0.104 | 0.075 | 1.39
ArraysSort.floatParallelSort | 50 | 0.277 | 0.215 | 1.29
ArraysSort.floatParallelSort | 75 | 0.466 | 0.243 | 1.92
ArraysSort.floatParallelSort | 100 | 0.587 | 0.322 | 1.82
ArraysSort.floatParallelSort | 1000 | 9.503 | 4.349 | 2.19
ArraysSort.floatParallelSort | 10000 | 534.867 | 214.857 | 2.49
ArraysSort.floatParallelSort | 100000 | 1028.12 | 1022.735 | 1.01
ArraysSort.floatParallelSort | 1000000 | 9537.956 | 4622.919 | 2.06
</body>

</html>

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (3 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 2 [Authors](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8319577](https://bugs.openjdk.org/browse/JDK-8319577): x86_64 AVX2 intrinsics for Arrays.sort methods (int, float arrays) (**Enhancement** - P4)

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**) ⚠️ Review applies to [bc590d9f](https://git.openjdk.org/jdk/pull/16534/files/bc590d9fecb49071c912bb1e194498db8ba5f21a)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to [bc590d9f](https://git.openjdk.org/jdk/pull/16534/files/bc590d9fecb49071c912bb1e194498db8ba5f21a)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16534/head:pull/16534` \
`$ git checkout pull/16534`

Update a local copy of the PR: \
`$ git checkout pull/16534` \
`$ git pull https://git.openjdk.org/jdk.git pull/16534/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16534`

View PR using the GUI difftool: \
`$ git pr show -t 16534`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16534.diff">https://git.openjdk.org/jdk/pull/16534.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16534#issuecomment-1811094775)

### CODE REVIEW

File: make/modules/java.base/Lib.gmk
Comment (vamsi-parasa) [reply make/modules/java.base/Lib.gmk:245]: Hi Magnus, the new x86-simd-sort 4.0 needs C++17 to compile. Will look into the changes needed for this library to compile without the C++17 standard and get back to you.

Thanks,
Vamsi

Comment (dholmes-ora) [reply make/modules/java.base/Lib.gmk:245]: Seems a bit odd to me too as the existing simd code seems to C code residing in .cpp files for some reason.

File: src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp
Comment (dholmes-ora) [inline src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp:3]: Is this person an OpenJDK Contributor?

Comment (robilad) [reply src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp:3]: Not listed here: https://oca.opensource.oracle.com/?ojr=contributors

Comment (sviswa7) [reply src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp:3]: Yes, Vamsi is part of Intel Java team. He also has the author status (https://openjdk.org/census#sparasa).

File: src/hotspot/share/opto/library_call.cpp
Comment (sviswa7) [inline src/hotspot/share/opto/library_call.cpp:5391]: UseAVX is platform specific and cannot be used in library_call.cpp.

Comment (vamsi-parasa) [reply src/hotspot/share/opto/library_call.cpp:5391]: Thanks Sandhya, will fix this issue.

File: src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp
Comment (dholmes-ora) [reply src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp:3]: @sviswa7 I am asking about the copyright holder "Serge Sans Paille".

Comment (vamsi-parasa) [reply src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp:3]: > @sviswa7 I am asking about the copyright holder "Serge Sans Paille".

Hi David (@dholmes-ora),

Sorry for the late reply as I was on vacation last week.

The SIMD sort for Java/OpenJDK is based on Intel's [x86-simd-sort](https://github.com/intel/x86-simd-sort) library. Serge Sans Paille (@serge-sans-paille) was one of the contributors to the x86-simd-sort. That is the reason his name is listed in the copyright notice. 

Thanks,
Vamsi

File: src/hotspot/share/opto/library_call.cpp
Comment (vamsi-parasa) [reply src/hotspot/share/opto/library_call.cpp]: Thanks Jatin for the suggestion. Please see your suggestion incorporated in the new code upstreamed.

Comment (vamsi-parasa) [reply src/hotspot/share/opto/library_call.cpp]: Thanks Sandhya for suggesting the change to use supports_simd_sort(BasicType bt). Please see the updated code upstreamed.

File: src/hotspot/share/opto/library_call.cpp
Comment (vamsi-parasa) [reply src/hotspot/share/opto/library_call.cpp]: Please see the checks moved to matcher_x86.hpp. Was suggested the same yesterday.

File: src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp
Comment (dholmes-ora) [reply src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp:3]: I take it he is not an Intel employee, in which case he has to be an OpenJDK contributor himself for code for which he holds a copyright to be contributed to OpenJDK. @robilad please correct me if I am wrong here.

Comment (serge-sans-paille) [reply src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp:3]: hey o/ No problem on my side to either let go my copyright or fill the contributor agreement (where is it?)

File: src/hotspot/share/opto/library_call.cpp
Comment (sviswa7) [reply src/hotspot/share/opto/library_call.cpp]: @vamsi-parasa Thanks, your changes look good to me.

File: make/modules/java.base/Lib.gmk
Comment (magicus) [reply make/modules/java.base/Lib.gmk:245]: Yeah, that is also a good point. Why setting C++ flags in CFLAGS, and not just CXXFLAGS?

Also @vamsi-parasa, did you check into if this really was needed and why? To be clear, I basically consider this an integration blocker.

Comment (vamsi-parasa) [reply make/modules/java.base/Lib.gmk:245]: Hi Magnus (@magicus),
The `-std=c++17` flag was to be placed in CXXFLAGS. Will fix that soon. 
Also, the latest version of x86-simd-sort is using C++17 features (which are supported from GCC 8 onwards). Would it be possible to restrict the availability of simdsort library in OpenJDK to GCC version >= 8 ?

Comment (magicus) [reply make/modules/java.base/Lib.gmk]: Raising the minimum gcc version is not done willy-nilly. (I feel a "You just don't ..." meme coming up)

But you are saying that you want to skip building this library unless you have a gcc version that supports c++17? 

I still don't really like it. I'd like to hear someone else who can think clearly about this, if we want to go down this path, and start adding libraries that use C++17. Maybe @kimbarrett has some input?

File: src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp
Comment (robilad) [reply src/java.base/linux/native/libsimdsort/avx2-32bit-qsort.hpp:3]: Thanks Serge! And thank you for sending in an OCA for processing, which has now been done.

File: make/modules/java.base/Lib.gmk
Comment (magicus) [reply make/modules/java.base/Lib.gmk]: Then you must add logic to check for this. Now the build will just fail if building with an older gcc. That is not acceptable.

Comment (vamsi-parasa) [reply make/modules/java.base/Lib.gmk]: Hi Marcus (@magicus), please see the updated code which added guards to check for GCC version >= 7.5 in src/java.base/linux/native/libsimdsort/{avx2-linux-qsort.cpp, avx512-linux-qsort.cpp}. Made sure that OpenJDK builds without errors using both GCC 7.5 and GCC 6.4.

**Review by sviswa7 — APPROVED** [review]
The PR looks good to me.

**Comment by vamsi-parasa** [comment]
Hello Vladimir (@vnkozlov),

Could you please review this PR?

Thanks,
Vamsi

File: make/modules/java.base/Lib.gmk
Comment (magicus) [reply make/modules/java.base/Lib.gmk]: That sounds weird. You can't check for if compiler options should be enabled or not inside source code files.

Are you saying that when compiling with GCC 6, it will just silently ignore `-std=c++17`? I'd have assumed that it printed a warning or error about an unknown or invalid option, if C++17 is not supported.

Comment (vamsi-parasa) [reply make/modules/java.base/Lib.gmk]: Hi Magnus (@magicus),
 
> Are you saying that when compiling with GCC 6, it will just silently ignore `-std=c++17`? I'd have assumed that it printed a warning or error about an unknown or invalid option, if C++17 is not supported.

The GCC complier for versions 6 (and even 5) silently ignores the flag `-std=c++17`. It does not print any warning or error. I tested it with a toy C++ program and also by building OpenJDK using GCC 6. 

> You can't check for if compiler options should be enabled or not inside source code files.

 what I meant was, there are #ifdef guards using predefined macros in the C++ source code to check for GCC version and make the simdsort code available for compilation or not based on the GCC version

```cpp
// src/java.base/linux/native/libsimdsort/simdsort-support.hpp
#if defined(_LP64) && (defined(__GNUC__) && ((__GNUC__ > 7) || ((__GNUC__ == 7) && (__GNUC_MINOR__ >= 5))))
        #define __SIMDSORT_SUPPORTED_LINUX
#endif
```

```cpp
//src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp
#include "simdsort-support.hpp"
#ifdef __SIMDSORT_SUPPORTED_LINUX
            <simdsort functions>
#endif
```

File: src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp
Comment (jatin-bhateja) [inline src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp]: Please add safe assertions for missing types.

File: src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp
Comment (jatin-bhateja) [inline src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp:234]: Can be removed if not being used.

File: src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp
Comment (jatin-bhateja) [inline src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp:277]: Can be removed if not being used.

File: src/java.base/linux/native/libsimdsort/avx512-32bit-qsort.hpp
Comment (jatin-bhateja) [inline src/java.base/linux/native/libsimdsort/avx512-32bit-qsort.hpp:235]: Can be removed.

File: src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp
Comment (jatin-bhateja) [inline src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp:56]: Please add assertion for unsupported types.

File: src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp
Comment (jatin-bhateja) [inline src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp:44]: Assertions for unsupported types.

File: src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp
Comment (jatin-bhateja) [inline src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp:64]: Lut64 is needed for compress64 emulation, can be removed.

File: make/modules/java.base/Lib.gmk
Comment (magicus) [reply make/modules/java.base/Lib.gmk]: Okay, then I guess I am fine with this.

**Review by magicus — APPROVED** [review]
Build changes look fine. You will still need the usual 2 reviewers from hotspot.

/reviewers 3

File: src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp
Comment (vamsi-parasa) [reply src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp]: Removed in the latest commit...

File: src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp
Comment (vamsi-parasa) [reply src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp:44]: Added in the latest commit...

File: src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp
Comment (vamsi-parasa) [reply src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp:58]: Removed in the latest commit...

File: src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp
Comment (vamsi-parasa) [reply src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp]: Removed in the latest commit...

File: src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp
Comment (vamsi-parasa) [reply src/java.base/linux/native/libsimdsort/avx2-emu-funcs.hpp]: Removed in the latest commit...

File: src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp
Comment (vamsi-parasa) [reply src/java.base/linux/native/libsimdsort/avx2-linux-qsort.cpp]: This is from an older (but outdated) commit. The assertions have been added in other cases.

File: make/modules/java.base/Lib.gmk
Comment (vamsi-parasa) [reply make/modules/java.base/Lib.gmk]: Thank you Magnus!

**Review by jatin-bhateja — APPROVED** [review]
LGTM, thanks!

**Comment by vamsi-parasa** [comment]
> LGTM, thanks!

Thanks Jatin!

**Comment by sviswa7** [comment]
@TobiHartmann @vnkozlov Please advice if we can go head and integrate this PR today before the fork.

**Comment by vnkozlov** [comment]
> @TobiHartmann @vnkozlov Please advice if we can go head and integrate this PR today before the fork.

Too late. Changes looks fine to me (I am still on fence that we moving to C++ implementation of intrinsics and require latest C++ compiler version). I need to run testing for latest version of changes before approval. Lets not rush.

**Comment by sviswa7** [comment]
Thanks a lot @vnkozlov, we will wait for your approval.

File: src/hotspot/share/opto/library_call.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/library_call.cpp:5393]: This check should be in `C2Compiler::is_intrinsic_supported()`

File: src/hotspot/share/opto/library_call.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/library_call.cpp:5450]: Same.

Comment (vamsi-parasa) [reply src/hotspot/share/opto/library_call.cpp]: Please see the updated changes which use C2Compiler::is_intrinsic_supported(id, bt)

**Comment by vamsi-parasa** [comment]
> @vamsi-parasa, sorry, I was wrong. I missed that you need to check type `bt`. Latest change is more complicated than it was before. Please revert it back (undo last change). I will test previous version 09.
@vnkozlov 
Vladimir, please see the commit reverted in the updated changes pushed now.

**Comment by vnkozlov** [comment]
Good. I submitted testing.

**Comment by vnkozlov** [comment]
@vamsi-parasa, please, remind me which tests check that code in `libsmdsort.so` is used?

**Comment by vamsi-parasa** [comment]
> @vamsi-parasa, please, remind me which tests check that code in `libsmdsort.so` is used?

@vnkozlov 
Please see the tests for simd sort code in `test/jdk/java/util/Arrays/Sorting.java`

**Comment by vnkozlov** [comment]
Testing have only one failure in closed tests and I need to fix it before this can be pushed.

**Comment by vamsi-parasa** [comment]
> Testing have only one failure in closed tests and I need to fix it before this can be pushed.

Thanks Vladimir for the update. Is the test failure because of this PR?

**Comment by vnkozlov** [comment]
> > Testing have only one failure in closed tests and I need to fix it before this can be pushed.
> 
> Thanks Vladimir for the update. Is the test failure because of this PR?

Yes. One of our test, which checks integrity of built JDK, is confused by changes in libsimdsort.so.

**Review by vnkozlov — APPROVED** [review]
I pushed closed changes.

**Comment by vamsi-parasa** [comment]
> I pushed closed changes.

Thanks Vladimir!

File: make/modules/java.base/Lib.gmk
Comment (magicus) [reply make/modules/java.base/Lib.gmk]: @vamsi-parasa You said:
> Made sure that OpenJDK builds without errors using both GCC 7.5 and GCC 6.4.

but now we have https://bugs.openjdk.org/browse/JDK-8321688. Did you introduce any changes after you tested with GCC 7.5? It seems strange to me that the code simultaneously both works and not works with gcc 7.5.

Comment (vamsi-parasa) [reply make/modules/java.base/Lib.gmk]: Hi Magnus (@magicus), did a fresh pull of the OpenJDK and was able to build it successfully (without any errors) using GCC 7.5.0 on Ubuntu Linux machine. 
(I am on vacation till Jan7th, 2024. Our team will look into this issue)

Comment (magicus) [reply make/modules/java.base/Lib.gmk]: New information in JDK-8321688 says it is only happening on slowdebug. This is probably why it was missed.

**Pull request closed without merging**
